modules/storages/app/contracts/storages/storages/base_contract.rb

```diff
@@ -0,0 +1,107 @@
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require 'net/http'
+require 'uri'
+
+# Purpose: common functionalities shared by CreateContract and UpdateContract
+# UpdateService by default checks if UpdateContract exists
+# and uses the contract to validate the model under consideration
+# (normally it's a model).
+module Storages::Storages
+  class BaseContract < ::ModelContract
+    MINIMAL_NEXTCLOUD_VERSION = 23
+
+    include ::Storages::Storages::Concerns::ManageStoragesGuarded
+    include ActiveModel::Validations
+
+    attribute :name
+    validates :name, length: { minimum: 1, maximum: 255 }, allow_nil: false
+
+    attribute :provider_type
+    validates :provider_type, inclusion: { in: ->(*) { Storages::Storage::PROVIDER_TYPES } }, allow_nil: false
+
+    attribute :creator, writable: false do
+      validate_creator_is_user
+    end
+
+    attribute :host
+    validates :host, length: { minimum: 1, maximum: 255 }, allow_nil: false
+    validates_url :host
+
+    # Check that a host actually is a storage server.
+    # But only do so if the validations above for URL were successful.
+    validate :validate_host_reachable, if: -> { errors[:host].empty? }
+
+    def validate_creator_is_user
+      unless creator == user
+        errors.add(:creator, :invalid)
+      end
+    end
+
+    def validate_host_reachable
+      response = request_capabilities
+
+      unless response.is_a? Net::HTTPSuccess
+        errors.add(:host, :invalid)
+        return
+      end
+
+      unless major_version_sufficient?(response)
+        errors.add(:host, :invalid)
+      end
+    end
+
+    def major_version_sufficient?(response)
+      return false unless response.body
+
+      version = JSON.parse(response.body).dig('ocs', 'data', 'version', 'major')
+      return false if version.nil?
+      return false if version < MINIMAL_NEXTCLOUD_VERSION
+
+      true
+    end
+
+    private
+
+    def request_capabilities
+      uri = URI.parse(File.join(host, '/ocs/v2.php/cloud/capabilities'))
+      request = Net::HTTP::Get.new(uri)
+      request["Ocs-Apirequest"] = "true"
+      request["Accept"] = "application/json"
+
+      req_options = {
+        use_ssl: uri.scheme == "https"
+      }
+
+      Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
+        http.request(request)
+      end
```

Comment:
`Net::HTTP::Get` has a default timeout of 60 seconds. That means that in the worst case, the validation will take 60s+. Even base case, it will not be fast given that there is an external request. Because of that, we might want to reduce the times this is triggered. We could do that be checking whether the `host` value changed and only validate if it did.

Speaking about external requests; Outgoing traffic might be prevented by the network OpenProject is placed in. Currently, the only outgoing request by OpenProject I can think of is to S3 in case that is configured to be used as a storage provider. We should add that to the list of infrastructure requirements.

Reply:
> `Net::HTTP::Get` has a default timeout of 60 seconds. That means that in the worst case, the validation will take 60s+. Even best case, it will not be fast given that there is an external request. Because of that, we might want to reduce the times this is triggered. We could do that be checking whether the `host` value changed and only validate if it did.

I set the open and read timeout to 5 and 3, and no retries. The check is done only if `host_changed?`. If changed, it will still do a request each time `valid?` is called on the contract (it does not cache the result). I did not do it because I assumed `valid` is called only once in the code. Is it ok?

Also, about the error codes, whether the host is unreachable or the Nextcloud version is too low, we do `errors.add(:host, :invalid)`, and the user has the message "Host is invalid". We can provide better error messages for both cases, saying that host is unreachable, or that the Nextcloud version is too low. What do you think @wielinde @ulferts 

> Speaking about external requests; Outgoing traffic might be prevented by the network OpenProject is placed in. Currently, the only outgoing request by OpenProject I can think of is to S3 in case that is configured to be used as a storage provider. We should add that to the list of infrastructure requirements.

If some people want to integrate with Nextcloud, I hope they'll allow outgoing requests to it.

Where can I amend this list of infrastructure requirements?

Reply:
Error messages improved in 8b6b19538a.